src/librustc_mir_build/hair/pattern/_match.rs

```diff
@@ -893,11 +893,12 @@ impl<'tcx> Constructor<'tcx> {
                                     // Treat all uninhabited types in non-exhaustive variants as
                                     // `TyErr`.
                                     (_, true, true) => cx.tcx.types.err,
-                                    // Treat all non-visible fields as `TyErr`. They can't appear
-                                    // in any other pattern from this match (because they are
+                                    // Treat all non-visible fields as `()`. They can't appear in
+                                    // any other pattern from this match (because they are
                                     // private), so their type does not matter - but we don't want
-                                    // to know they are uninhabited.
-                                    (false, ..) => cx.tcx.types.err,
+                                    // to know they are uninhabited, so we choose an inhabited
+                                    // dummy.
+                                    (false, ..) => cx.tcx.types.unit,
```

Comment:
Maybe the ICE is because this has a constructor with `0` leaves? You could try some type like `u8` I guess.
But I wonder if `ty::Error` is specially-handled elsewhere in this module.